sophie-kaleba/assignment1/Anagram.java

```diff
@@ -1,86 +1,90 @@
  */
 public class Anagram {
-	
+
 	/**
-	* Check if the two arguments given as parameter are anagrams. Handle both case sensitive and case insensitive comparisons.
-	* example: areAnagrams(listen, silent, false) returns True, areAnagrams(Listen, silent, true) returns False, areAnagrams(apple, pabble, true) returns False
-	* @param element1 - the 1st element
-	* @param element2 - the 2d element
-	* @param caseSenstive - true if the letter case is important ("listen" != "Listen")
-	* @return true if the 2 elements are anagrams, false otherwise
-	*
-	*/
+	 * Check if the two arguments given as parameter are anagrams. 
+	 * Handle both case sensitive and case insensitive comparisons.
+	 * example: areAnagrams(listen, silent, false) returns True, 
+	 * example: areAnagrams(Listen, silent, true) returns False
+	 * example: areAnagrams(apple, pabble, true) returns False
+	 * @param element1 - the 1st element
+	 * @param element2 - the 2d element
+	 * @param caseSenstive - true if the letter case is important ("listen" != "Listen")
+	 * @return true if the 2 elements are anagrams, false otherwise
+	 *
+	 */
 	public static boolean areAnagrams(String element1, String element2, boolean caseSensitive) {
-	    /* the element can be a sentence, split it according the standard control characters*/
+		/* the element can be a sentence, split it according the standard control characters*/
 		String[] splittedElement1 = element1.split(",|\\;| |\\:|\\.");
 		String[] splittedElement2 = element2.split(",|\\;| |\\:|\\.");
-		
+
 		/* if the element returned by split contains at least 2 words, this is a sentence */
 		if (splittedElement1.length>1 || splittedElement2.length>1 ) {
-		    return Anagram.AreAnagramsSentences(splittedElement1, splittedElement2);
+			return Anagram.AreAnagramsSentences(splittedElement1, splittedElement2, caseSensitive);
 		}
 		return Anagram.AreAnagramsWords(splittedElement1[0], splittedElement2[0], caseSensitive);
 	}
 
-	
+
 	/**
 	 * compares 2 sentences and check if they are anagrams
 	 * @param sentence1 - the 1st sentence
 	 * @param sentence2 - the 2d sentence
 	 * @return true if the 2 sentences are anagrams, false otherwise
 	 */
-	public static boolean AreAnagramsSentences(String[] sentence1, String[] sentence2) {
-		List<String> s1 = Arrays.asList(sentence1);
-		List<String> s2 = Arrays.asList(sentence2);
-		
+	public static boolean AreAnagramsSentences(String[] sentence1, String[] sentence2, boolean caseSensitive) {
+
 		/* must contains the same amount of words to be anagrams */
-		if (s1.size() == s2.size()) {
+		if (sentence1.length == sentence2.length) {
 			for (String word : sentence1) {
-				int i = 0;
 				boolean found = false;
-				while (!found) {
-					if (i==s2.size()) return false; /* the anagram of this word is not in the resulting sentence */
-					found = Anagram.AreAnagramsWords(word, s2.get(i), false);
-					i++;
+				for (String word2 : sentence2) {
+					if (AreAnagramsWords(word, word2, caseSensitive)){
```

Comment:
As implemented, this would allow different words from the first sentence, to be matched against the same word in the second sentence.

This can lead to an error for two sentences like:

```java
  areAnagrams("a b b c", "a b c c")
```

Can we write a test to capture this case?
How can we fix this issue?

Reply:
Indeed! I just added the test case.
One solution could consist in deleting the found word in the 2d sentence, but that would have a negative impact on complexity (both in terms of space and time), at least for a naive implementation (O(n) for iterating through the array to find the word, and the space for an additional array containing one less element).
Another solution could consist in storing the "found" indexes in a data structure and ignoring them during the comparison phase. That adds a test, and an additional data structure.
The second option seems better imo, what do you think ?